pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/DateTruncTransformFunctionTest.java

```diff
@@ -0,0 +1,172 @@
+import java.nio.file.Paths;
+import java.time.ZoneOffset;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.pinot.core.operator.transform.function.DateTruncTransformFunction.DateTimeZoneIndex.getDateTimeZone;
+import static org.apache.pinot.core.operator.transform.function.TimeZoneKey.UTC_KEY;
+
+public class DateTruncTransformFunctionTest
+{
+  private static final String TIME_COLUMN = "time";
+  private static final ZoneOffset WEIRD_ZONE = ZoneOffset.ofHoursMinutes(7, 9);
+  private static final DateTimeZone WEIRD_DATE_TIME_ZONE = DateTimeZone.forID(WEIRD_ZONE.getId());
+  private static final DateTime WEIRD_TIMESTAMP = new DateTime(2001, 8, 22, 3, 4, 5, 321, WEIRD_DATE_TIME_ZONE);
+  private static final String WEIRD_TIMESTAMP_ISO8601_STRING = "2001-08-22T03:04:05.321+07:09";
+  private static final DateTimeZone UTC_TIME_ZONE = getDateTimeZone(UTC_KEY);
+  private static final String TIMESTAMP_ISO8601_STRING = "2001-08-22T03:04:05.321+00:00";
+
+  private static final DateTime TIMESTAMP = new DateTime(2001, 8, 22, 3, 4, 5, 321, UTC_TIME_ZONE); // This is TIMESTAMP w/o TZ
+
+
+  private static long iso8601ToUtcEpochMillis(String iso8601) {
+    DateTimeFormatter formatter = ISODateTimeFormat.dateTimeParser().withOffsetParsed();
+    return formatter.parseDateTime(iso8601).getMillis();
+  }
+
+  private static void testDateTruncHelper(String literalInput, String unit, String tz, long expected) throws Exception {
+    long zmillisInput = iso8601ToUtcEpochMillis(literalInput);
+    GenericRow row = new GenericRow();
+    row.init(ImmutableMap.of(TIME_COLUMN, zmillisInput));
+    List<GenericRow> rows = ImmutableList.of(row);
+    Schema schema = new Schema();
+    schema.addField(new TimeFieldSpec(TIME_COLUMN, FieldSpec.DataType.LONG, TimeUnit.MILLISECONDS));
+
+    SegmentGeneratorConfig config = new SegmentGeneratorConfig(schema);
+    String segmentName = "testSegment";
+    String indexDirPath = Paths.get(Files.createTempDirectory("pinot_date_trunc_test").toAbsolutePath().toString(), segmentName).toAbsolutePath().toString();
+    try {
+      FileUtils.deleteQuietly(new File(indexDirPath));
+      config.setOutDir(indexDirPath);
+      config.setSegmentName(segmentName);
+      SegmentIndexCreationDriverImpl driver = new SegmentIndexCreationDriverImpl();
+      driver.init(config, new GenericRowRecordReader(rows, schema));
+      driver.build();
+
+      IndexSegment indexSegment = ImmutableSegmentLoader.load(new File(indexDirPath, segmentName), ReadMode.heap);
+      Set<String> columnNames = indexSegment.getPhysicalColumnNames();
+      HashMap<String, DataSource> dataSourceMap = new HashMap<>(columnNames.size());
+      for (String columnName : columnNames) {
+        dataSourceMap.put(columnName, indexSegment.getDataSource(columnName));
+      }
+
+      ProjectionBlock projectionBlock = new ProjectionOperator(dataSourceMap,
```

Comment:
I think it would be better to test this transform function using overall query execution code path as well in addition to this test. Here you have to build the expression tree, operator, projection block etc and this is something that query execution code already does. So may be we should also write a test extending BaseQueriesTest similar to how other query execution tests are written.

Reply:
That way we will know that this transform function fits into query execution code flow

Reply:
Thanks for your comments and looking over this diff.

I am wondering if there is an existing test framework/example I can "copy" ? I think it is a bit onerous for UDF authors to develop whole new testing frameworks. So if you can point me in the right direction with what I should base this test on, I will be happy to author it and amend the diff.

Reply:
Yes, please look at TransformQueriesTest and how it extends BaseQueriesTest. The test:

(1) Creates a sample data set
(2) Build segment(s) out of the dataset
(3) Runs inner segment level query
(4) Runs inter segment (and inter server) query.

Probably (4) is not necessary for this test. 

You can also look at DistintQueriesTest.java in the same directory as TransformQueriesTest.